onnxruntime/python/onnxruntime_inference_collection.py

```diff
@@ -331,44 +335,51 @@ def clear_binding_inputs(self):
     def clear_binding_outputs(self):
         self._iobinding.clear_binding_outputs()
 
+    _numpy_obj_references = []
 
 class OrtValue:
     '''
     This class provides APIs to construct and deal with OrtValues.
     '''
-    def __init__(self, ortvalue):
+    def __init__(self, ortvalue, numpy_obj = None):
         if isinstance(ortvalue, C.OrtValue):
             self._ortvalue = ortvalue
+            # Hold a ref count to the numpy object if the OrtValue is backed directly
+            # by its data buffer so that it isn't destroyed when the OrtValue is in use
+            self._numpy_obj = numpy_obj
         else:
             # An end user won't hit this error
             raise ValueError("`Provided ortvalue` needs to be of type " +
                              "`onnxruntime.capi.onnxruntime_pybind11_state.OrtValue`")
 
     @staticmethod
-    def tensor_from_numpy(numpy_obj, device_type='cpu', device_id=0):
+    def ortvalue_from_numpy(numpy_obj, device_type='cpu', device_id=0):
         '''
         Factory method to construct an OrtValue (which holds a Tensor) from a given Numpy object
-        A copy of the data in the Numpy object is held by the OrtValue
-        :param numpy_obj: The Numpy object to construct the Tensor from
+        A copy of the data in the Numpy object is held by the OrtValue only if the device is NOT cpu
+        :param numpy_obj: The Numpy object to construct the OrtValue from
         :param device_type: e.g. cpu, cuda, cpu by default
         :param device_id: device id, e.g. 0
         '''
-        return OrtValue(C.OrtValue.tensor_from_numpy(numpy_obj, C.OrtDevice(get_ort_device_type(device_type),
-                        C.OrtDevice.default_memory(), device_id)))
+        # Hold a reference to the numpy object (if device_type is 'cpu') as the OrtValue
+        # is backed directly by the data buffer of the numpy object and so the numpy object
+        # must be around until this OrtValue instance is around
+        return OrtValue(C.OrtValue.ortvalue_from_numpy(numpy_obj, C.OrtDevice(get_ort_device_type(device_type),
```

Comment:
same as above- hold onto a reference if the device is cpu as the OrtValue will directly use the numpy object's buffer